common4j/src/main/com/microsoft/identity/common/java/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java

```diff
@@ -716,13 +727,21 @@ private String getCloudSpecificTokenEndpoint(final MicrosoftAuthorizationRespons
         return buildCloudSpecificTokenEndpoint((MicrosoftStsAuthorizationResponse) response);
     }
 
+    private void setAuthenticationScheme(@NonNull final AbstractAuthenticationScheme authScheme) {
+        mAuthScheme = authScheme;
+    }
+
     /**
      * Gets the at/pop device credential's thumbprint.
      *
      * @return The at/pop device credential thumbprint.
      */
     @Nullable
     public String getDeviceAtPopThumbprint() {
+        if (mAuthScheme instanceof PopAuthenticationSchemeWithClientKeyInternal) {
+           return ((PopAuthenticationSchemeWithClientKeyInternal) mAuthScheme).getKid();
+        }
+
```

Comment:
Rather then having this `if/else` block now to handle get thumbprint for 2 different types of auth scheme, we can move the getThumbprint method directly to the `IPopAuthSchemeParams` interface and implement it over there. As a hint, `mPopManager` is already present inside `PopAuthenticationSchemeInternal` class so what you need is already there. Once you do that then the code here becomes very simple, you can just invoke:

`mAuthScheme.getKeyThumbprint()` and you wouldn't have to worry about the specific type of pop auth scheme at this layer.

Reply:
Good suggestion,
Currently IPopAuthSchemeParams  is also implemented on the Msal side in PopAuthenticationScheme class, which doesn't have PopManager.
We can either remove that class and start using authscheme classes from common, (it will require some refactoring and would be a breaking change) or we can keep the getKeyThumbprint as no-op in that class.
For now, I am keeping that as out of scope for current change